### PR TITLE
Port to glib 0.14

### PR DESCRIPTION
**Pull request by cgwalters**
Port to glib 0.14

This depends on the WIP port for ostree-rs.

---

Update to use new GLib variant bindings

This drops some unsafe code and increases ergonomics.  I'd like
to kill our `variant_utils` entirely but that will still need
more work.

---

### CODE REVIEW

**Comment by cgwalters** [comment]
Depends https://github.com/ostreedev/ostree-rs/pull/7

**Comment by cgwalters** [comment]
OK published new https://crates.io/crates/ostree 0.12 and switched to use that here!

File: lib/src/variant_utils.rs [hunk 1]
```diff
@@ -32,36 +32,21 @@ pub fn variant_normal_from_bytes(ty: &str, bytes: glib::Bytes) -> glib::Variant
...
-        let r = glib_sys::g_variant_new_fixed_array(
-            b"y\0".as_ptr() as *const _,
-            buf.as_ptr() as *const _,
-            buf.len(),
-            size_of::<u8>(),
-        );
-        glib_sys::g_variant_ref_sink(r);
-        from_glib_full(r)
-    }
+    buf.to_variant()
```
Comment (cgwalters) [inline lib/src/variant_utils.rs:35]: Being able to use stuff like this was a primary motivation for the bump - look ma no `unsafe`!

Comment (cgwalters) [reply lib/src/variant_utils.rs]: And done!  Almost all of `variant_utils.rs` is gone, and the only `Variant` bits are just `pub(crate)`.  There's still `VariantDictExt` which I think is kind of still useful...needs some investigation.

File: lib/src/ima.rs [hunk 2]
```diff
@@ -47,10 +48,19 @@ fn xattrs_to_map(v: &glib::Variant) -> BTreeMap<Vec<u8>, Vec<u8>> {
     map
 }
 
-/// Reserialize a map to GVariant of type `a(ayay)`
-fn xattrmap_serialize(map: &BTreeMap<Vec<u8>, Vec<u8>>) -> glib::Variant {
-    let map: Vec<_> = map.iter().collect();
-    variant_utils::new_variant_a_ayay(&map)
+/// Create a new GVariant of type a(ayay).  This is used by OSTree's extended attributes.
+fn new_variant_a_ayay<'a, T: 'a + AsRef<[u8]>>(
```
Comment (cgwalters) [inline lib/src/ima.rs:52]: Not gonna lie after I wrote this I felt like I'd just inscribed a new page in a spellbook.

File: lib/src/tar/import.rs [hunk 3]
```diff
@@ -434,7 +433,7 @@ impl<'a> Importer<'a> {
         let mut contents = vec![0u8; n as usize];
         entry.read_exact(contents.as_mut_slice())?;
         let contents: glib::Bytes = contents.as_slice().into();
-        let contents = variant_new_from_bytes(OSTREE_XATTRS_FORMAT, contents, false);
+        let contents = Variant::from_bytes::<&[(&[u8], &[u8])]>(&contents);
```
Comment (cgwalters) [inline lib/src/tar/import.rs:436]: Note how we were previously passing a string `a(ayay)` and now encode the Variant type via Rust generics.

File: lib/src/ima.rs [hunk 4]
```diff
@@ -47,10 +48,19 @@ fn xattrs_to_map(v: &glib::Variant) -> BTreeMap<Vec<u8>, Vec<u8>> {
...
+) -> glib::Variant {
+    let children: Vec<_> = items
+        .into_iter()
+        .map(|(a, b)| {
+            let a = a.as_ref();
+            let b = b.as_ref();
+            Variant::from_tuple(&[a.to_variant(), b.to_variant()])
+        })
+        .collect();
+    Variant::from_array::<(&[u8], &[u8])>(&children)
```
Comment (lucab) [inline lib/src/ima.rs:63]: Self-note: this internally gets ownership of the content, see note at https://developer.gnome.org/glib/stable/glib-GVariant.html#g-variant-new-array.

**Review by lucab — APPROVED** [review]

**Pull request merged**